### PR TITLE
Visualisation of MCMC iteration/time series 

### PR DESCRIPTION
**Pull request by gordonkoehn**
Building upon #38, here we are building the actual plots. 

The envisioned product. 

- [x] 1) The plot with the X-axis being the iteration number, and Y axis being the loglikelihood of the sampled tree.
- [x] 2) The number of iterations needed to sample the true tree. - to some JSON/ txt
- [x] 3) Plot of the distance to the true tree on the Y axis for several chains and the iteration number on the X axis.

This PR is the 2/2 for closing issue #23.

### CODE REVIEW

**Comment by gordonkoehn** [comment]
The first version of the plots:

dist_prob_over_iterations

TODO: 

- [ ] add txt if true tree was reached
- [ ] plot initial and final tree / most likely trees

**Comment by gordonkoehn** [comment]
mmmh... this all seems very cumbersome to pice together the mcmc logP and distances. 

An mcmc run should be analyzed and it's data saved to a separate the an `data/mcmc/mark00/analytics.csv` which contain the columns: 

`iteration | is_true_tree | logP | similarity:MP3 | similarity:example2` 

This should be done by one script. Followed by the visualisation, which may then plot a single or multiple MCMc runs together.

**Comment by gordonkoehn** [comment]
Alright, I'll keep this PR small as it is. I'll do the following before I submit for review:

Add the plotting of the 
- [ ] log_prob
- [ ] distance
individually. 

- [ ] Save the distances to file in NumPy array with a column of "is_true_tree"
- [ ] consider integrating `DistMcmcData`class as well as `is_true_tree` 

**the separation of analysis, i.e. calculating distances and identifying trees and saving, and the plotting, i.e. reading in the easily readable data, excluding trees and plotting them, is to be done in the next PR**

File: scripts/plot_trees.py [hunk 1]
```diff
@@ -101,32 +101,15 @@ def main() -> None:
...
-
-        save_name = "iter_" + str(i) + "_log_prob_" + str(log_prob)
-
-        # make full path with pathlib
+
         save_dir = Path(args.out_dir)
 
-        # plot tree
-        visualize.plot(tree, save_name, save_dir, print_options)
+        visualize.plot_tree_mcmc_sample(sample, save_dir)
```
Comment (gordonkoehn) [inline scripts/plot_trees.py:112]: refactoring code out to specialized functions

File: scripts/run_mcmc.py [hunk 2]
```diff
@@ -196,8 +159,13 @@ def run_chain(
...
 
-    # load observed mutation matrix data from file
-    mut_mat = jnp.array(get_mutation_matrix(params.data_fp))
+    # load data of mutation matrix
+    with open(params.data_fp, "r") as f:
+        cell_simulation_data = json.load(f)
+    cell_simulation_data = tree_inf.get_simulation_data(cell_simulation_data)
+    # get the mutation matrix
+    # TODO: adjust to flexible noisy / perfect
+    mut_mat = cell_simulation_data["perfect_mutation_mat"]
```
Comment (gordonkoehn) [inline scripts/run_mcmc.py:168]: all changes above are refactoring code to more general functions

File: src/pyggdrasil/distances/_interface.py [hunk 3]
```diff
@@ -7,8 +7,8 @@
 _IntegerTreeRoot: TypeAlias = TreeNode[int, Any]
 
 
-class TreeSimilarity(Protocol):
-    """Interface for similarity functions between the trees.
+class TreeSimilarityMeasure(Protocol):
```
Comment (gordonkoehn) [inline src/pyggdrasil/distances/_interface.py:10]: I wanted to have once common type or both `TreeDistance` and `TreeSimilarity` -  let me know what you think of the design.

File: src/pyggdrasil/interface/__init__.py [hunk 4]
```diff
@@ -76,3 +80,71 @@ def append(self, iteration: int, tree: TreeNode, log_probability: float):
         self.iterations = jnp.append(self.iterations, iteration)
         self.trees.append(tree)
         self.log_probabilities = jnp.append(self.log_probabilities, log_probability)
+
+
+@dataclass
+class AugmentedMcmcData(PureMcmcData):
+    """Distance/Similarity and True Tree boolean enhanced MCMC data
+
+    intended to be human-readable and easy to use for plotting"""
```
Comment (gordonkoehn) [inline src/pyggdrasil/interface/__init__.py:89]: This class is not used currently. I tried to come up with a good format to save all MCMC data and analytics, excluding trees. Frankly, a pandas data frame might have been easier. I am planning to write functions to convert from `PureMcmcData` to this `AugmentenMcmcData`, by making simple choices for which Distances to use and then save this to disk as a csv or dataframe. 

Once that works, I'll refactor the visualization scripts to read in these files - to separate the analytics from the visualisation. Perhaps even creating a new submodule `analyze` ? 

What do you think ? I am not very convinced by my design of this class.

File: src/pyggdrasil/interface/__init__.py [hunk 5]
```diff
@@ -76,3 +80,71 @@ def append(self, iteration: int, tree: TreeNode, log_probability: float):
...
+            self.true_tree[iteration],
+        )
+
+    def append(
+        self,
+        iteration: int,
+        tree: TreeNode,
+        log_probability: float,
+        similarity_scores: Union[dict[TreeSimilarityMeasure, float]] = None,
+        true_tree: Optional[bool] = None,
```
Comment (gordonkoehn) [inline src/pyggdrasil/interface/__init__.py:125]: Am having a lot of trouble with overwriting / extending the function of the child class here. - Think I am missing something here. Think what I don't get is https://en.wikipedia.org/wiki/Liskov_substitution_principle - Pyright is giving me a hard time when I am assigning a `None` to a dictionary. What is a good way to make a dictionary optional ?

File: src/pyggdrasil/visualize/__init__.py [hunk 6]
```diff
@@ -1,5 +1,23 @@
...
+)
+
+__all__ = [
+    "plot_tree",
+    "make_mcmc_run_panel",
+    "save_dist_iteration",
+    "save_log_p_iteration",
+    "plot_tree_mcmc_sample",
+    "plot_tree_no_print",
+]
```
Comment (gordonkoehn) [inline src/pyggdrasil/visualize/__init__.py:23]: Created a lot of public API, that I had missed before

File: src/pyggdrasil/visualize/_tree.py [hunk 7]
```diff
@@ -172,3 +169,52 @@ def plot(
     plt.savefig(fullpath + ".svg", bbox_inches="tight")
     plt.close()
     logger.info(f"Saved tree plot to {fullpath}.svg")
+
+
```
Comment (gordonkoehn) [inline src/pyggdrasil/visualize/_tree.py:173]: Below: specialized plotting functions, that define arguments for the most general plotting function automatically. `plot_tree`

File: workflows/Snakefile [hunk 8]
```diff
@@ -125,5 +126,72 @@ rule run_mcmc_mark00:
             --out_dir {params.out_dir}
         """
```
Comment (gordonkoehn) [inline workflows/Snakefile:128]: Below: two messy snakemake rules to run the two scripts created. Will rewrite the snakemake completely soon. I'll think about a sensible way to store all data and plots then.

**Review by gordonkoehn — COMMENTED** [review]
Some unexpected groundwork and tinkering with writing modular plotting functions. Single MCMC run visualizations work - see below. An additional script to get the true tree from the simulation data. Quite a bit of refactoring.

Presenting a potential class to store all MCMC run data and distances together. Will be bound together in the next PR, where an mcmc run shall be analyzed and saved to disk. Adhoc calculation of distances will re refactored then.

Not done: check run for true tree

dist_iter
dist_logP_iter
logP_iter
top_tree_1
true_tree

**Review by gordonkoehn — COMMENTED** [review]
Some unexpected groundwork and tinkering with writing modular plotting functions. Single MCMC run visualizations work - see below. An additional script to get the true tree from the simulation data. Quite a bit of refactoring.

Presenting a potential class to store all MCMC run data and distances together. Will be bound together in the next PR, where an mcmc run shall be analyzed and saved to disk. Adhoc calculation of distances will re refactored then.

Not done: check run for true tree

dist_iter
dist_logP_iter
logP_iter
top_tree_1
true_tree

**Comment by gordonkoehn** [comment]
Reason for current test failure: `AugmentedMcmcData` class has some type issues with overwriting the `append` function of the BaseClass.

**Comment by gordonkoehn** [comment]
@pawel-czyz could you please have a brief look at this somewhat updated PR.  It integrates a few functions and classes I am referencing in all later PRs. I'll probably change quite a bit once I redo the visualisation to work from multiple runs and separate the calculation of distances from the visualization. 

Some of the built errors are new pydantic errors - after an update. And some minor import problems.

**Review by pawel-czyz — COMMENTED** [review]
I've taken a look and added some comments on the questions – let me know if they are sensible or not :sweat_smile:

**Review by pawel-czyz — APPROVED** [review]
